src/Workspaces/Core/Portable/Options/EditorConfig/EditorConfigDocumentOptionsProviderFactory.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorLogger;
+
+namespace Microsoft.CodeAnalysis.Options.EditorConfig
+{
+    [Export(typeof(IDocumentOptionsProviderFactory)), Shared]
+    internal sealed class EditorConfigDocumentOptionsProviderFactory : IDocumentOptionsProviderFactory
+    {
+        public IDocumentOptionsProvider Create(Workspace workspace)
+        {
+            return new EditorConfigDocumentOptionsProvider(workspace.Services.GetService<IErrorLoggerService>());
+        }
+
+        private class EditorConfigDocumentOptionsProvider : IDocumentOptionsProvider
+        {
+            private readonly IErrorLoggerService _errorLogger;
+
+            public EditorConfigDocumentOptionsProvider(IErrorLoggerService errorLogger)
+            {
+                _errorLogger = errorLogger;
+            }
+
+            public async Task<IDocumentOptions> GetOptionsForDocumentAsync(Document document, CancellationToken cancellationToken)
+            {
+                var options = await document.GetAnalyzerOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+                return new DocumentOptions(options, _errorLogger);
+            }
+
+            private class DocumentOptions : IDocumentOptions
+            {
+                private readonly ImmutableDictionary<string, string> _options;
+                private readonly IErrorLoggerService _errorLogger;
+
+                public DocumentOptions(ImmutableDictionary<string, string> options, IErrorLoggerService errorLogger)
+                {
+                    _options = options;
+                    _errorLogger = errorLogger;
+                }
+
+                public bool TryGetDocumentOption(OptionKey option, OptionSet underlyingOptions, out object value)
+                {
+                    var editorConfigPersistence = option.Option.StorageLocations.OfType<IEditorConfigStorageLocation>().SingleOrDefault();
```

Comment:
i hate SingleOrDefault

Reply:
Yeah, me too. This is [following the pattern from the legacy provider](https://github.com/dotnet/roslyn/blob/e6fed01a172644d1dbe4a780d027069c8b752beb/src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.DocumentOptions.cs#L25).